### PR TITLE
Parvezp/add methods

### PR DESCRIPTION
**Pull request by ppathan**
# Type of change? 

- [ ] Bug fix
- [x] New feature
- [ ] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
Enabling methods for the new simulation brancx

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: Services/DeviceMethods.cs [hunk 1]
```diff
@@ -136,7 +139,8 @@ private void MethodExecution(MethodRequest methodRequest)
                 this.scriptInterpreter.Invoke(
                     this.cloudToDeviceMethods[methodRequest.Name],
                     scriptContext,
-                    this.deviceState);
+                    this.deviceState,
+                    this.deviceProperties);
```
Comment (timlaverty) [inline Services/DeviceMethods.cs]: I like this split out now - makes the code clearer.

File: Services/DeviceMethods.cs [hunk 2]
```diff
@@ -136,7 +139,8 @@ private void MethodExecution(MethodRequest methodRequest)
                 this.scriptInterpreter.Invoke(
                     this.cloudToDeviceMethods[methodRequest.Name],
                     scriptContext,
-                    this.deviceState);
+                    this.deviceState,
```
Comment (timlaverty) [inline Services/DeviceMethods.cs]: Now that you're passing properties separately, do you want to add them to the log.Debug above?
i.e. this:

this.log.Debug("Executing method for device", () => new
                 {	                 {
                     this.deviceId,	                     this.deviceId,
                     methodName = methodRequest.Name,	                     methodName = methodRequest.Name,
                     this.deviceState	                     this.deviceState
                 });	                 });

Comment (ppathan) [reply Services/DeviceMethods.cs]: Fixed

File: Services/DeviceClient.cs [hunk 3]
```diff
@@ -185,5 +187,10 @@ private static TwinCollection SmartDictionaryToTwinCollection(ISmartDictionary d
 
             return result;
         }
+
+        public Task UpdateTwinAsync(Device device)
```
Comment (jillcary) [inline Services/DeviceClient.cs]: This was renamed to "UpdatePropertiesAsync" in the `send-twin-updates` branch.

File: Services/data/devicemodels/scripts/EmergencyValveRelease-method.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Entry point function called by the simulation engine.
+ *
+ * @param context        The context contains current time, device model and id
+ * @param previousState  The device state since the last iteration
+ */
+
+/*jslint unparam: true*/
+function main(context, previousState) {
```
Comment (jillcary) [inline Services/data/devicemodels/scripts/EmergencyValveRelease-method.js]: The latest scripts include "previousProperties" as a parameter and have a more descriptive comment for the main() function. Should this also match that style?

File: Services/data/devicemodels/scripts/FirmwareUpdate-method.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+/**
+ * Entry point function called by the simulation engine.
+ *
+ * @param context        The context contains current time, device model and id, not used
+ * @param previousState  The device state since the last iteration, not used
+ */
+/*jslint unparam: true*/
+function main(context, previousState) {
```
Comment (jillcary) [inline Services/data/devicemodels/scripts/FirmwareUpdate-method.js]: Should this also include previousProperties?

File: SimulationAgent/DeviceConnection/DeviceConnectionActor.cs [hunk 6]
```diff
@@ -77,6 +79,9 @@ public enum ActorEvents
         /// </summary>
         private IDeviceStateActor deviceStateActor;
 
+        public ISmartDictionary DeviceState => this.deviceStateActor.DeviceState;
```
Comment (jillcary) [inline SimulationAgent/DeviceConnection/DeviceConnectionActor.cs]: Nit: Consider adding comments to match the style of the other members

File: Services/DeviceClient.cs [hunk 7]
```diff
@@ -19,8 +19,9 @@ public interface IDeviceClient
         Task ConnectAsync();
         Task DisconnectAsync();
         Task SendMessageAsync(string message, DeviceModel.DeviceModelMessageSchema schema);
-        Task UpdatePropertiesAsync(ISmartDictionary properties);
-        Task RegisterMethodsForDeviceAsync(IDictionary<string, Script> methods, Dictionary<string, object> deviceState);
+        Task UpdateTwinAsync(Device device);
+        Task RegisterMethodsForDeviceAsync(IDictionary<string, Script> methods, ISmartDictionary deviceState, ISmartDictionary deviceProperties);
```
Comment (jillcary) [inline Services/DeviceClient.cs]: Are there any tests for the new methods functionality?

**Pull request merged**